### ISSUE TITLE
Prevent flashing HTML when submitting CKEditor forms

### ISSUE THREAD

**Issue by swampopus** [backdrop/backdrop-issues#7149]
## Description of the bug

When editing any content type that uses CKEditor5, when you click 'Save', the editor is destroyed and you can see a flash of plain HTML where the editor used to be.  This looks funky and confuses less-savvy users.

See video which demonstrates the behavior on a clean install:

https://github.com/user-attachments/assets/e828b83d-5524-4ba9-86e5-a9b518b40d9d

## Steps To Reproduce

To reproduce the behavior:
1. Create a clean install of Backdrop (optional)
2. Click to "Create Page"
3. Make sure the editor is set to Basic, and using CKEditor for the Body field.
4. Type anything into the Body field
5. Save, and watch closely, you will see a flash of HTML.

## Actual behavior

See video above.

## Solution!

There's no reason to destroy the editor `(line 123 of core/modules.ckeditor5/js/ckeditor5.js)` if the form is being submitted anyway.

I have a kludge/hack that actually fixes this; will submit a PR shortly.  Basically, I set a variable if I know the form is submitting, then, in ckeditor5.js, I only call `editor.destroy()` if we are NOT submitting the form.  This works in my case, but I don't know if it works in all situations.

## Additional information

Add any other information that could help, such as:
- Backdrop CMS version:  1.34.0
- Web server and its version: Apache 2.4.65
- PHP version: 8.1.34
- Database sever (MySQL or MariaDB?) and its version: MariaDB 10.7.8
- Operating System and its version: Debian
- Browser(s) and their versions: Chrome 148

**Comment by swampopus** [comment]
PR is ready for testing

**Comment by olafgrabienski** [comment]
I can confirm the issue, and the PR fixes it for me.

**Comment by swampopus** [comment]
Note:  Until this PR is merged (if accepted), I have created a workaround contrib module: https://github.com/backdrop-contrib/ckeditor_no_html_flash

**Comment by herbdool** [comment]
The approach seems  okay to me.

**Comment by quicksketch** [comment]
@swampopus I reviewed your PR and I'm not thrilled about using a global variable for this purpose. I submitted an alternative PR that makes a new parameter on `Backdrop.filterEditorDetach()`: https://github.com/backdrop/backdrop/pull/5299 Could you check if that approach looks acceptable?

This may also be useful for TinyMCE or other editors that may want to distinguish between when they need to clean up their memory and event bindings and when they can just let the browser handle the cleanup. It would be great if @indigoxela could weigh in here.

**Comment by swampopus** [comment]
@quicksketch, I think your approach (adding the new parameter) can definitely work. 

I considered that too-- but I didn't want to add a parameter (I was afraid it might break something unforseen), which is why I went with adding a global var instead, which seemed like the least amount of risk.

Also, I imagined there may be some future hypothetical module that wants to know when the editor's form is being submitted, not having to do with detaching the editor.

I have a similar issue posted already for TinyMCE (https://github.com/backdrop-contrib/tinymce/issues/198).  I am also curious to hear from @indigoxela.

Correct me if I'm wrong, but if TinyMCE's `detach` function does _not_ get the new parameter because the user hasn't upgraded, attempting to pass it a 4th param isn't going to cause any errors to show in the console is it?  Likewise, if the user has an older version of Backdrop but a newer version of TinyMCE with a 4th parameter expected, and nothing is passed to it, this also won't cause an error to show in the console?

What I'm getting at is as long as no error shows in the browser console, then I don't see any reason adding the extra parameter to the `detach` functions wouldn't work.

**Comment by quicksketch** [comment]
> Correct me if I'm wrong, but if TinyMCE's detach function does not get the new parameter because the user hasn't upgraded, attempting to pass it a 4th param isn't going to cause any errors to show in the console is it?

I don't think JS complains about this when it happens, but it might vary based on whether `"use strict"` is set or not.

Here's another approach that I also think may be safe: introducing another option for the existing "trigger" parameter. In theory, `filter.js` is supposed to support "unload", "serialize", and "move" values for the "trigger" parameter, but in practice, it is only using "serialize" and `void`. If we fix `filter.js` to use "unload" like `ajax.js`, then we also have the opportunity to add another option for "submit" (or whatever we want to call it). I filed another PR at https://github.com/backdrop/backdrop/pull/5300 that uses this approach. I think it should be safe for all existing editor modules but we should confirm, especially with tinyMCE.

**Comment by indigoxela** [comment]
Wow, OK, 3 possibilities to choose from... 😆 

JS doesn't "complain" if one doesn't use a function param. And whether it's there or not can be checked via "undefined". So, that should work.

Trigger "submit" seems like the most logical approach to me, but requires some considerations. It would work with Tiny integration, as it does no explicit check for "unload", but... 🤔 Need to think that over (newer core + older tiny / newer tiny + older core...).

**Comment by swampopus** [comment]
Thanks for the clarification re: javascript.

@quicksketch -- I like the "trigger" variable approach best so far.  I was hoping when all this began that `trigger` would tell me something like "what element triggered this action?" or something similar, and I could use it for that.  

But, every time I checked it, it was always `undefined`.  So I think using `trigger`  makes the most semantic sense, and it doesn't add any new parameters or global vars.

**Comment by quicksketch** [comment]
We all seem to like the approach in https://github.com/backdrop/backdrop/pull/5300 the most so I've closed the other two PRs.

I fixed the conditional and @swampopus has already verified it. @swampopus if you agree, you could move this to RTBC.

**Comment by swampopus** [comment]
@quicksketch It all looks good to me-- feel free to proceed

**Comment by herbdool** [comment]
It's not clear yet if this last remaining PR has been tested as reviewed. @swampopus have you tested it? 

@quicksketch my knowledge here is shallow but the code looks good to me.

**Comment by swampopus** [comment]
@herbdool Yes, I have tested it.  I just re-tested, and made a new content type with "unlimited" editor fields, and had no issues changing formats, adding new fields, etc., and it still behaved correctly when saving.

**Comment by herbdool** [comment]
@swampopus thanks. I've updated the labels so it's clear.

**Comment by laryn** [comment]
Thanks @swampopus for filing the issue and a PR for consideration as well as reviewing, and to @quicksketch for reviewing and filing TWO PRs. Thanks also to @olafgrabienski, @indigoxela, and @herbdool for testing and code reviews. Merged into `1.x` and `1.34.x`.

**Comment by swampopus** [comment]
Awesome-- my first thing to get into core :)  (not my PR, but still, feels nice!)

**Comment by quicksketch** [comment]
It was a great idea and something that had bothered me quite a lot with CKEditor. Thanks @swampopus!

**Comment by swampopus** [comment]
Thanks!
